src/librustc/lint/levels.rs

```diff
@@ -470,6 +472,32 @@ impl<'a> LintLevelsBuilder<'a> {
         }
     }
 
+    /// If we have e.g. `#[allow(future_compat_lint)]` this will have
+    /// no effect as `min_level > Allow`. We want to tell the user about this.
+    fn lint_higher_minimum_attr_lint(
+        &self,
+        lint: &Lint,
+        level: Level,
+        src: &LintSource,
+        specs: &FxHashMap<LintId, (Level, LintSource)>,
+    ) {
+        let min_level = lint.min_level;
+        if min_level <= level {
+            return;
+        }
+
+        if let LintSource::Node(name, _, span, _) = src {
+            let lint = builtin::UNUSED_ATTRIBUTES;
+            let (lvl, src) = self.sets.get_lint_level(lint, self.cur, Some(&specs), &self.sess);
+            let msg = format!("#[{}({})] has no effect", level.as_str(), name);
+            lint::struct_lint_level(self.sess, lint, lvl, src, Some((*span).into()), &msg)
+                .note(&format!("the minimum lint level for `{}` is `{}`", name, min_level.as_str()))
+                .note(&format!("the lint level cannot be reduced to `{}`", level.as_str()))
+                .help("remove it")
```

Comment:
While I do smile at the simplicity of this command to the user, I worry that the "it" in the message is somewhat open to interpretation.

Maybe emit the slightly expanded: "remove the `#[{}({})]` directive"?

Reply:
Hmm; so what if the user wrote `#[allow(foo, bar)]` but only `foo` is problematic? The user didn't write `#[allow(foo)] #[allow(bar)]` in this case... Maybe that's OK?

Reply:
yeah, I was privately musing about where this could go wrong.

I guess ideally we would highlight span of the whole the whole directive if there is only one lint being allowed there, and just highlight the single name within the directive if there is more than one name within it?

Reply:
Hmm... so the current message results in roughly:
```
error: #[allow(illegal_floating_point_literal_pattern)] has no effect
  --> $DIR/lint-min-allow-no-effect.rs:4:10
   |
LL | #![allow(illegal_floating_point_literal_pattern)]
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/lint-min-allow-no-effect.rs:1:9
   |
LL | #![deny(unused_attributes)]
   |         ^^^^^^^^^^^^^^^^^
   = note: the minimum lint level for `illegal_floating_point_literal_pattern` is `warn`
   = note: the lint level cannot be reduced to `allow`
   = help: remove it
```

I don't think I have access to the span for the whole attribute but if we say "remove the `#[{}({})]` directive" and the `illegal_floating_point_literal_pattern` is underlined as above, then I think the user would understand what is meant from the overall context so I'm inclined to go with your original proposal here.